compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+      if (f.isDirectory) f.listFiles.flatMap(flatten)
+      else Array(f)
+
+    // Directories in which to compile all containing files with `flags`:
+    val targetDir = new JFile(outDir + "/" + sourceDir.getName + "/")
+    targetDir.mkdirs()
+
+    val target = ConcurrentCompilationTarget(flatten(sourceDir), flags, targetDir)
+    new CompilationTest(target)
+  }
+
+  /** Compiles all `files` together as a single compilation run. It is given a
+   *  `testName` since files can be in separate directories and or be otherwise
+   *  dissociated
+   */
+  def compileList(testName: String, files: List[String], flags: Array[String])(implicit outDirectory: String): CompilationTest = {
+    val outDir = outDirectory + getCallingMethod + "/" + testName + "/"
+
+    // Directories in which to compile all containing files with `flags`:
+    val targetDir = new JFile(outDir)
+    targetDir.mkdirs()
+    assert(targetDir.exists, s"couldn't create target directory: $targetDir")
+
+    val target = ConcurrentCompilationTarget(files.map(new JFile(_)).toArray, flags, targetDir)
+
+    // Create a CompilationTest and let the user decide whether to execute a pos or a neg test
+    new CompilationTest(target)
+  }
+
+  /** This function compiles the files and folders contained within directory
+   *  `f` in a specific way.
+   *
+   *  - Each file is compiled separately as a single compilation run
+   *  - Each directory is compiled as a `SeparateCompilationTaret`, in this
+   *    target all files are grouped according to the file suffix `_X` where `X`
+   *    is a number. These groups are then ordered in ascending order based on
+   *    the value of `X` and each group is compiled one after the other.
+   *
+   *  For this function to work as expected, we use the same convention for
+   *  directory layout as the old partest. That is:
+   *
+   *  - Single files can have an associated check-file with the same name (but
+   *    with file extension `.check`)
+   *  - Directories can have an associated check-file, where the check file has
+   *    the same name as the directory (with the file extension `.check`)
+   */
+  def compileFilesInDir(f: String, flags: Array[String])(implicit outDirectory: String): CompilationTest = {
+    val outDir = outDirectory + getCallingMethod + "/"
+    val sourceDir = new JFile(f)
+    checkRequirements(f, sourceDir, outDir)
+
+    val (dirs, files) = compilationTargets(sourceDir)
+
+    val targets =
+      files.map(f => ConcurrentCompilationTarget(Array(f), flags, toCompilerDirFromFile(f, sourceDir, outDir))) ++
+      dirs.map(dir => SeparateCompilationTarget(dir, flags, toCompilerDirFromDir(dir, sourceDir, outDir)))
+
+    // Create a CompilationTest and let the user decide whether to execute a pos or a neg test
+    new CompilationTest(targets)
+  }
+
+  /** This function behaves similar to `compileFilesInDir` but it ignores
+   *  sub-directories and as such, does **not** perform separate compilation
```

Comment:
Because we have tests that compile things on a single depth to weed out where an error ocurred - see `tastyBootstrap`